Translate Tutorial 1 of 2 (#33)
reactjs/pt-BR.reactjs.org: content/tutorial/tutorial.md

```bash
npx create-react-app my-app
```

3. Apague todos os arquivos na pasta `src/` do novo projeto

> Nota: **não exclua a pasta `src` inteira, apenas os arquivos originais dentro dela.** Substituiremos os arquivos de origem padrão por exemplos deste projeto na próxima etapa.

```bash
cd my-app
cd src

# Se você estiver usando um Mac ou Linux:
rm -f *

# Ou, se você estiver no Windows:
del *

# E então, volte para a pasta do projeto
cd ..
```

4. Adicione um arquivo chamado `index.css` na pasta` src/`com [este código CSS](https://codepen.io/gaearon/pen/oWWQNa?editors=0100).

5. Adicione um arquivo chamado `index.js` na pasta` src/`com [este código JS](https://codepen.io/gaearon/pen/oWWQNa?editors=0010).

6. Adicione estas três linhas ao topo do `index.js` na pasta `src/`:

```js
import React from 'react';
import ReactDOM from 'react-dom';
import './index.css';
```

Agora, se você executar o `npm start` na pasta do projeto e abrir `http://localhost:3000` no navegador, você verá um campo vazio de jogo da velha.

Recomendamos seguir [estas instruções](https://babeljs.io/docs/editors/) para configurar o realce de sintaxe para seu editor.

</details>

> ### Ajuda, estou preso! {#help-im-stuck}

Author: Não sei que palavra usar, o termo que me vem a cabeça seria empacado ou emperrado. Mas acho que isso é regional e nem todo mundo vai entender.

Reviewer: Acho que pode ser 
'Ajuda, estou com dificuldades!'

  - "docs/tutorial-zh-CN.html"
---

Este tutorial não assume nenhum conhecimento existente do React.

## Antes de começarmos o tutorial {#before-we-start-the-tutorial}

> Vamos construir um pequeno jogo durante este tutorial. **Você pode ficar tentado a ignorá-lo porque você não vai construir jogos - mas dê uma chance.** As técnicas que você aprenderá no tutorial são fundamentais para criar qualquer aplicativo React, e o domínio lhe dará uma chance. profundo entendimento de React.

Reviewer 2: ```suggestion
Vamos construir um pequeno jogo durante este tutorial. **Você pode ficar tentado a ignorá-lo porque você não vai construir jogos - mas dê uma chance.** As técnicas que você aprenderá no tutorial são fundamentais para criar qualquer aplicativo React e o domínio dará a você uma chance e um profundo entendimento de React.
```

> Vamos construir um pequeno jogo durante este tutorial. **Você pode ficar tentado a ignorá-lo porque você não vai construir jogos - mas dê uma chance.** As técnicas que você aprenderá no tutorial são fundamentais para criar qualquer aplicativo React, e o domínio lhe dará uma chance. profundo entendimento de React.

Author: 👍🏽

> Dica
>
> Este tutorial foi criado para pessoas que preferem ** aprender fazendo **. Se você preferir aprender conceitos do zero, confira nosso [step-by-step guide](/docs/hello-world.html). Você pode achar este tutorial e o guia complementares um ao outro.

O tutorial está dividido em várias seções:

* [Configuração do Tutorial](#setup-for-the-tutorial) lhe dará **um ponto de partida** para seguir o tutorial.
* [Visão geral](#overview) ensinará **os fundamentos** de React: componentes, propriedades e estado.
* [Completando o Jogo](#completing-the-game) will teach you **the most common techniques** in React development.
* [Adicionando Time Travel (viagem no tempo)](#adding-time-travel) lhe dará **uma visão mais profunda** para os pontos fortes exclusivos do React.

> Você não precisa completar todas as seções de uma vez para obter o valor deste tutorial. Tente chegar o mais longe possível, mesmo que seja uma ou duas seções.

Author: 👍🏽

> Não há problema em copiar e colar o código enquanto você acompanha o tutorial, mas recomendamos que você o digite à mão. Isso ajudará você a desenvolver uma memória muscular e um entendimento mais forte.

Reviewer 2: ```suggestion
Não há problema em copiar e colar o código enquanto você acompanha o tutorial. Mas, recomendamos que você o digite à mão. Isso ajudará você a desenvolver uma memória muscular e ter um entendimento mais forte.
```

### O que estamos construindo? {#what-are-we-building}

Neste tutorial, mostraremos como criar um jogo interativo de jogo-da-velha com React.

> Você pode ver o que vamos construir aqui: **[Resultado Final](https://codepen.io/gaearon/pen/gWWZgR?editors=0010)**. Se o código não fizer sentido para você, ou se você não estiver familiarizado com a sintaxe do código, não se preocupe! O objetivo deste tutorial é ajudá-lo a entender React e sua sintaxe.

Reviewer 2: ```suggestion
Você pode ver o que vamos construir aqui: **[Resultado Final](https://codepen.io/gaearon/pen/gWWZgR?editors=0010)**. Se o código não fizer sentido para você ou se você não estiver familiarizado com a sintaxe do código, não se preocupe! O objetivo deste tutorial é ajudá você a entender o React e sua sintaxe.
```

Author: 2e3083d9215fca43c44c2ccae014ace107f944dc

> Recomendamos que você confira o jogo tic-tac-toe antes de continuar com o tutorial. Uma das características que você notará é que existe uma lista numerada à direita do tabuleiro do jogo. Esta lista fornece um histórico de todas as jogadas que ocorreram no jogo e é atualizada à medida que o jogo avança.

Reviewer 2: ```suggestion
Recomendamos que você confira o jogo tic-tac-toe (jogo da velha) antes de continuar com o tutorial. Uma das características que você notará é que existe uma lista numerada à direita do tabuleiro do jogo. Esta lista fornece um histórico de todas as jogadas que ocorreram no jogo e é atualizada à medida que o jogo avança.
```

> Você pode fechar o jogo jogo da velha assim que estiver familiarizado com ele. Começaremos a partir de um modelo mais simples neste tutorial. Nosso próximo passo é prepará-lo para que você possa começar a construir o jogo.

Reviewer 2: ```suggestion
Você pode fechar o jogo da velha assim que estiver familiarizado com ele. Começaremos a partir de um modelo mais simples neste tutorial. Nosso próximo passo é prepará-lo para que você possa começar a desenvolver o jogo.
```

### Pré-requisitos {#prerequisites}

> Vamos supor que você tenha alguma familiaridade com HTML e JavaScript, mas você deve ser capaz de acompanhá-lo mesmo que esteja vindo de uma linguagem de programação diferente. Também vamos supor que você esteja familiarizado com conceitos de programação como funções, objetos, matrizes e, em menor escala, classes.

Reviewer 2: ```suggestion
Vamos presumir que você já tenha alguma familiaridade com HTML e JavaScript. Porém, você deve ser capaz de acompanhá-lo mesmo que esteja vindo de uma linguagem de programação diferente. Também vamos presumir que você também já esteja familiarizado com conceitos de programação. Tais como: funções, objetos, matrizes em menor escala e classes.
```

> Se você precisa revisar JavaScript, recomendamos ler [este guia](https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/A_re-introduction_to_JavaScript). Observe que também estamos usando alguns recursos do ES6 - uma versão recente do JavaScript. Neste tutorial, estamos usando [arrow function](https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Functions/Arrow_functions), [classes](https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Classes), [`let`](https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Statements/let), e declarações [`const`](https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Statements/const). Você pode usar o [Babel REPL] (babel://es5-syntax-example) para verificar para qual código o ES6 compila.

Reviewer 2: ```suggestion
Se você precisa fazer revisão sobre JavaScript, recomendamos que leia [este guia](https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/A_re-introduction_to_JavaScript). Observe que também estamos usando alguns recursos do ES6 - uma versão recente do JavaScript. Neste tutorial, estamos usando [arrow function](https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Functions/Arrow_functions), [classes](https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Classes), [`let`](https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Statements/let), e declarações [`const`](https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Statements/const). Você pode usar o [Babel REPL] (babel://es5-syntax-example) para verificar para qual código o ES6 compila.
```

Author: 👍🏽

## Configuração para o tutorial {#setup-for-the-tutorial}

Há duas maneiras de concluir este tutorial: você pode escrever o código em seu navegador ou configurar um ambiente de desenvolvimento local em seu computador.

> ### Opção de Configuração 1: Escrever Código no Navegador {#setup-option-1-write-code-in-the-browser}

Reviewer 2: ```suggestion
### Opção de Configuração 1: Desenvolvendo o Código no Navegador {#setup-option-1-write-code-in-the-browser}
```

Esta é a maneira mais rápida de começar!

> Primeiro, abra este **[Código Inicial](https://codepen.io/gaearon/pen/oWWQNa?editors=0010)** em uma nova guia. A nova aba deve exibir um tabuleiro de jogo tic-tac-toe vazio e o código React. Nós estaremos editando o código React neste tutorial.

Reviewer 2: ```suggestion
Primeiro, abra este **[Código Inicial](https://codepen.io/gaearon/pen/oWWQNa?editors=0010)** em uma nova guia. A nova aba deve exibir um tabuleiro de jogo tic-tac-toe (jogo da velha) vazio e o código React. Nós estaremos editando o código React neste tutorial.
```

Agora você pode ignorar a segunda opção de configuração e ir para a seção [Visão geral](#overview) para obter uma visão geral do React.

### Opção de instalação 2: ambiente de desenvolvimento local {#setup-option-2-local-development-environment}

Isto é completamente opcional e não é necessário para este tutorial!

<br>

<details>

<summary><b>Opcional: instruções para serem seguidas na sua máquina usando seu editor de texto preferido</b></summary>

> This setup requires more work but allows you to complete the tutorial using an editor of your choice. Here are the steps to follow:

Reviewer 3: @eduardomoroni Está faltando a tradução desse trecho

> This setup requires more work but allows you to complete the tutorial using an editor of your choice. Here are the steps to follow:

Author: 👍🏽

1. Certifique-se de ter uma versão recente do [Node.js](https://nodejs.org/pt-br/) instalada.
2. Siga as [instruções de instalação do creat-react-app](/docs/create-a-new-react-app.html#create-react-app) para criar um novo projeto.

```bash
npx create-react-app my-app
```

3. Apague todos os arquivos na pasta `src/` do novo projeto

> > Nota: **não exclua a pasta `src` inteira, apenas os arquivos originais dentro dela.** Substituiremos os arquivos de origem padrão por exemplos deste projeto na próxima etapa.

Reviewer 2: ```suggestion
 Nota: **não exclua a pasta `src` inteira. Apenas os arquivos originais dentro dela.** Substituiremos os arquivos de origem padrão por exemplos deste projeto na próxima etapa.
```

Author: https://github.com/reactjs/pt-BR.reactjs.org/pull/33/commits/a5b6ea93b11b53a7ed7de3a296a4ba6c53653f31

```bash
cd my-app
cd src

# Se você estiver usando um Mac ou Linux:
rm -f *

# Ou, se você estiver no Windows:
del *

# E então, volte para a pasta do projeto
cd ..
```

4. Adicione um arquivo chamado `index.css` na pasta` src/`com [este código CSS](https://codepen.io/gaearon/pen/oWWQNa?editors=0100).

5. Adicione um arquivo chamado `index.js` na pasta` src/`com [este código JS](https://codepen.io/gaearon/pen/oWWQNa?editors=0010).

6. Adicione estas três linhas ao topo do `index.js` na pasta `src/`:

```js
import React from 'react';
import ReactDOM from 'react-dom';
import './index.css';
```

Agora, se você executar o `npm start` na pasta do projeto e abrir `http://localhost:3000` no navegador, você verá um campo vazio de jogo da velha.

Recomendamos seguir [estas instruções](https://babeljs.io/docs/editors/) para configurar o realce de sintaxe para seu editor.

</details>

> ### Ajuda, estou com dificuldades! {#help-im-stuck}

Reviewer 2: @eduardomoroni acho q fica mais lindo assim. Não precisam fazer tradução ao pé da letra. ;)

```suggestion
### Me Ajudem. Estou com Dificuldades! {#help-im-stuck}
```

Você pode fechar o jogo jogo da velha assim que estiver familiarizado com ele. Começaremos a partir de um modelo mais simples neste tutorial. Nosso próximo passo é prepará-lo para que você possa começar a construir o jogo.

### Pré-requisitos {#prerequisites}

Vamos supor que você tenha alguma familiaridade com HTML e JavaScript, mas você deve ser capaz de acompanhá-lo mesmo que esteja vindo de uma linguagem de programação diferente. Também vamos supor que você esteja familiarizado com conceitos de programação como funções, objetos, matrizes e, em menor escala, classes.

> Se você precisa revisar JavaScript, recomendamos ler [este guia](https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/A_re-introduction_to_JavaScript). Observe que também estamos usando alguns recursos do ES6 - uma versão recente do JavaScript. Neste tutorial, estamos usando [arrow function](https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Functions/Arrow_functions), [classes](https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Classes), [`let`](https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Statements/let), e declarações [`const`](https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Statements/const). Você pode usar o [Babel REPL] (babel://es5-syntax-example) para verificar para qual código o ES6 compila.

Reviewer 4: Na última linha, o espaço (indicado por  `␣`) não está de acordo com a sintaxe do MD: `[Babel REPL]␣(babel://es5-syntax-example)`

Reviewer 5: PR com o ajuste, obrigado @micalevisk

  }
}

// Exemplo de uso: <ShoppingList name="Mark" />
```

> Nós falaremos sobre tags às tags engraçadas em formato XML em breve. Usamos componentes para dizer ao React o que queremos ver na tela. Quando nossos dados forem alterados, o React atualizará e renderizará novamente com eficiência nossos componentes.

Reviewer 2: ```suggestion
Nós falaremos sobre formatos engraçados de tags que possuem formato de XML em breve. Usamos componentes para dizer ao React o que queremos ver na tela. Quando nossos dados forem alterados, o React atualizará e renderizará novamente com eficiência nossos componentes.
```

> Nós falaremos sobre tags às tags engraçadas em formato XML em breve. Usamos componentes para dizer ao React o que queremos ver na tela. Quando nossos dados forem alterados, o React atualizará e renderizará novamente com eficiência nossos componentes.

Author: 👍🏽

> Aqui, o ShoppingList é um **componente React de classe** ou **component React do tipo classe**. Um componente recebe parâmetros, chamados `props` (abreviação de propriedades), e retorna uma hierarquia de views para exibir através do método `render`.

Author: 👍🏽

> O método `render` retorna uma *descrição* do que você deseja ver na tela. React pega a descrição e exibe o resultado. Em particular, `render` retorna um **elemento React **, que é uma descrição leve do que renderizar. A maioria dos desenvolvedores do React usa uma sintaxe especial chamada "JSX", que facilita a gravação dessas estruturas. A sintaxe `<div />` é transformada em tempo de compilação para `React.createElement ('div')`. O exemplo acima é equivalente a:

Author: 👍🏽

```javascript
return React.createElement('div', {className: 'shopping-list'},
  React.createElement('h1', /* ... filhos de h1 ... */),
  React.createElement('ul', /* ... filhos de ul ... */)
);
```

[Veja a versão completa completa.](babel://tutorial-expanded-version)

Se você está curioso, o `createElement()` é descrito em mais detalhes na [referência da API](/docs/react-api.html#createelement), mas não iremos usá-lo neste tutorial. Em vez disso, continuaremos usando o JSX.

> O JSX vem com todo o poder do JavaScript. Você pode colocar *qualquer* expressões JavaScript dentro de chaves dentro do JSX. Cada elemento React é um objeto JavaScript que você pode armazenar em uma variável ou passar em seu programa.

Author: 👍🏽

O componente `ShoppingList` acima apenas renderiza componentes internos do DOM como `<div />` e ` <li />`. Mas você também pode compor e renderizar componentes React personalizados. Por exemplo, agora podemos nos referir a toda a lista de compras escrevendo `<ShoppingList />`. Cada componente React é encapsulado e pode operar de forma independente; Isso permite que você construa interfaces complexas a partir de componentes simples.

## Inspecting the Starter Code {#inspecting-the-starter-code}

Se você for trabalhar no tutorial **em seu navegador,** abra esse código em uma nova guia: **[Código Inicial](https://codepen.io/gaearon/pen/oWWQNa?editors=0010 )** Se você vai trabalhar no tutorial **localmente,** abra `src/index.js` em sua pasta de projeto (você já tocou este arquivo durante a [configuração](#setup-option-2-local-development-environment)).

> Este Código Inicial é a base do que estamos construindo. Fornecemos o estilo CSS para que você só precise se concentrar no aprendizado do React e na programação do jogo do jogo da velha.

Author: 👍🏽

Ao inspecionar o código, você notará que temos três componentes React:

> * Quadrado(Square)

Author: Eu pensei em traduzir completamento os termos no decorrer da tradução, porém os código exemplo vão estar em Ingles. O post faz referencia de um link externo do CodePen em que as pessoas vão poder olhar o código. Se eu traduzir todos os termos teriamos que traduzir o nome dos componentes no código exemplo, o que eu não acho que seja a melhor ideia.

>}
>```
>
>Note que com `onClick = {() => alert ('click')}`, estamos passando *uma função* como prop `onClick`. Ela Só é acionada após um clique. Esquecer o `() =>` e escrever somente `onClick = {alert ('click')}` é um erro comum, e dispararia o alerta toda vez que o componente fosse renderizado novamente.

> Como próximo passo, queremos que o componente Square "lembre" que foi clicado e preencha com um "X". Para "lembrar" as coisas, os componentes usam o **estado**.

Reviewer 3: ```suggestion
Como próximo passo, queremos que o componente Square "lembre" que foi clicado e preencha com um "X". Para "lembrar" as coisas, os componentes usam o **estado (_state_)**.
```

> Os componentes React podem ter estado configurando `this.state` em seus construtores. `this.state` deve ser considerado como privado para o componente React que o definiu. Vamos armazenar o valor atual do Square em `this.state` e alterá-lo quando o Square for clicado.

Reviewer 3: ```suggestion
Os componentes React podem ter estado (_state_) configurando `this.state` em seus construtores. `this.state` deve ser considerado como privado para o componente React que o definiu. Vamos armazenar o valor atual do Square em `this.state` e alterá-lo quando o Square for clicado.
```

}
```

>Nota
>

> >Em [classes JavaScript](https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Classes), você sempre precisa chamar `super` ao definir o construtor de uma subclasse. Todas os componente de classe React que possuem um `construtor` devem iniciá-lo com uma chamada `super (props)`.

Reviewer 3: ```suggestion
>Em [classes JavaScript](https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Classes), você sempre precisa chamar `super` ao definir o construtor de uma subclasse. Todas os componentes de classe React que possuem um método `constructor` devem iniciá-lo com uma chamada `super (props)`.
```

Como texto entre aspas invertidas representa código, acredito que não devemos traduzir `constructor` nesse caso. Sugeri adicionar a descrição de que é um método para ficar igual como na documentação do mozilla que foi linkada :)

> Agora vamos mudar o método `render` do componente Square para exibir o valor do estado atual quando clicado:

Reviewer 3: ```suggestion
Agora vamos mudar o método `render` do componente Square para exibir o valor do estado (_state_) atual quando clicado:
```

* Substitua `this.props.value` por` this.state.value` dentro da tag `<button>`.

> * Substitua o manipulador de eventos `() => alert ()` com `() => this.setState ({value: 'X'})`.

Reviewer 3: ```suggestion
* Substitua o manipulador de eventos `() => alert ()` por `() => this.setState ({value: 'X'})`.
```

}
```

> Chamando `this.setState` a partir de um manipulador `onClick` no método `render` do componente Square, nós dizemos ao React para renderizar novamente aquele Square sempre que seu` <button>` for clicado. Após a atualização, o `this.state.value` do Square será `'X'`, então vamos ver o `X` no tabuleiro do jogo. Se você clicar em qualquer quadrado, um `X` deve aparecer.

Reviewer 3: ```suggestion
Ao chamar `this.setState` a partir de um manipulador `onClick` no método `render` do componente Square, nós dizemos ao React para renderizar novamente aquele Square sempre que seu` <button>` for clicado. Após a atualização, o `this.state.value` do Square será `'X'`, então vamos ver o `X` no tabuleiro do jogo. Se você clicar em qualquer quadrado, um `X` deve aparecer.
```

Quando você chama `setState` em um componente, o React atualiza automaticamente os componentes filhos dentro dele também.

**[Ver o código completo até este momento](https://codepen.io/gaearon/pen/VbbVLg?editors=0010)**

### Developer Tools {#developer-tools}

> A extensão React Devtools para [Chrome](https://chrome.google.com/webstore/detail/react-developer-tools/fmkadmapgofadopljbjfkapdkoienihi?hl=en) e [Firefox](https://addons.mozilla.org/en-US/firefox/addon/react-devtools/) permite inspecionar uma árvore de componentes  React com as ferramentas de desenvolvedor do seu navegador.

Reviewer 3: ```suggestion
A extensão React Devtools para [Chrome](https://chrome.google.com/webstore/detail/react-developer-tools/fmkadmapgofadopljbjfkapdkoienihi?hl=en) e [Firefox](https://addons.mozilla.org/en-US/firefox/addon/react-devtools/) permite inspecionar uma árvore de componentes React com as ferramentas de desenvolvedor do seu navegador.
```

<img src="../images/tutorial/devtools.png" alt="React Devtools" style="max-width: 100%">

> O React DevTools permite que você verifique as props e o estado de seus componentes React.

Reviewer 3: ```suggestion
O React DevTools permite que você verifique as props e o estado (_state_) de seus componentes React.
```

> Depois de instalar o React DevTools, você pode clicar com o botão direito do mouse em qualquer elemento da página, clicar em "Inspecionar" para abrir as ferramentas do desenvolvedor, e a guia React aparecerá como a última guia à direita.

Reviewer 3: ```suggestion
Depois de instalar o React DevTools, você pode clicar com o botão direito do mouse em qualquer elemento da página, clicar em "Inspecionar" para abrir as ferramentas de desenvolvedor, e a guia React aparecerá como a última guia à direita.
```

Quando você chama `setState` em um componente, o React atualiza automaticamente os componentes filhos dentro dele também.

**[Ver o código completo até este momento](https://codepen.io/gaearon/pen/VbbVLg?editors=0010)**

### Developer Tools {#developer-tools}

A extensão React Devtools para [Chrome](https://chrome.google.com/webstore/detail/react-developer-tools/fmkadmapgofadopljbjfkapdkoienihi?hl=en) e [Firefox](https://addons.mozilla.org/en-US/firefox/addon/react-devtools/) permite inspecionar uma árvore de componentes React com as ferramentas de desenvolvedor do seu navegador.

<img src="../images/tutorial/devtools.png" alt="React Devtools" style="max-width: 100%">

O React DevTools permite que você verifique as props e o estado (_state_) de seus componentes React.

Depois de instalar o React DevTools, você pode clicar com o botão direito do mouse em qualquer elemento da página, clicar em "Inspecionar" para abrir as ferramentas de desenvolvedor, e a guia React aparecerá como a última guia à direita.

> ** No entanto, observe que há algumas etapas extras para a extensão funcionar com o CodePen: **

Reviewer 4: Os espaçamentos antes e depois dos `**` não estão de acordo com a [sintaxe do MD](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#emphasis).

Alô @jhonmike xD

Reviewer 5: Ajustado no seguinte PR: https://github.com/reactjs/pt-BR.reactjs.org/pull/229 aguardando revisão
